### ISSUE TITLE
a11y: Router should automate focus management

### ISSUE THREAD

**Issue by ewinslow** [angular/angular#6106]
Proper focus management can really improve user experience for everyone, but is something that is easy to forget during development. It would be awesome if Angular did some smart things by default. Here are my suggestions:

When navigating to a new page:
 * if [autofocus] is set on any element, focus that element
 * otherwise focus the component itself (set tabindex to -1 if necessary)

When navigating back
 * restore focus to the element that was focused, if such info is available
 * otherwise, fall back to above rules

**Comment by bennadel** [comment]
I just stumbled across the fact that `autofocus` on an input doesn't seem to work when navigating to a page via a router / user interaction. I agree that this just "work."

**Comment by AlmeroSteyn** [comment]
Really awesome when people are concerned about a11y!

I agree that the `autofocus` should work on a total page load to work as specified. 

In general, however, taking over focus is something that should be used carefully. I expect autofocussing the component itself could cause more a11y issues than it tries to solve:

- It would make assumptions for what the user would want to do.
- If the app is hosted inside a page section (for example using Angular inside a CMS) it would autobypass the rest of the page function and render things like skiplinks obsolete (in fact this could already happen with nested components). 
- If you have nested `router outlets` which component should focus?
- By setting `tabindex=-1` on the component element and focussing this, it would force users to fully decorate each and every component with ARIA (i,e. `aria-describedby`).

**Comment by vicb** [comment]
If this is still an issue with the new router, please re-open an issue referencing the new router. thanks.

**Comment by snebjorn** [comment]
This is still an issue.

Here's a demonstration of the issue https://stackblitz.com/edit/angular-jvfn5g
